scipy/optimize/_quasi_newton_approx.py

```diff
@@ -0,0 +1,261 @@
+"""quasi-Newton Hessian approximations"""
+
+from __future__ import division, print_function, absolute_import
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import get_blas_funcs
+
+
+class QuasiNewtonApprox(object):
+    """Quasi-Newton approximation."""
+
+    _syr = get_blas_funcs('syr', dtype='d')  # Symetric rank 1 update
+    _syr2 = get_blas_funcs('syr2', dtype='d')  # Symetric rank 2 update
+    _symv = get_blas_funcs('symv', dtype='d')  # Symetric matrix-vector product
+
+    def instanciate_matrix(self, delta_x, delta_grad, approx_type='hess'):
+        """Instanciate unscaled matrix."""
+        n = len(delta_x)
+        self.approx_type = approx_type
+        if approx_type not in ('hess', 'inv_hess'):
+            raise ValueError("Unknown approx_type.")
+        # Create matrix
+        if self.approx_type == 'hess':
+            self.B = np.eye(n, dtype=float)
+        else:
+            self.H = np.eye(n, dtype=float)
+
+    def scale_matrix(self, delta_x, delta_grad):
+        """Scale matrix.
+
+        Heuristic to scale matrix at first iteration described
+        in Nocedal and Wright "Numerical Optimization"
+        p.143 formula (6.20)
+        """
+        y = delta_grad
+        s = delta_x
+        y_norm2 = np.dot(y, y)
+        ys = np.dot(y, s)
+        if ys == 0.0:
```

Comment:
Really minor, but it can be rewritten shorter:

```
if ys == 0:
    return

if self.approx_type == 'hess':
    self.B *= y_norm2 / ys
else:
    self.H = ys / y_norm2
```

Reply:
Also, it looks like it should be different testing for different `self.approx_type`. Like `ys == 0` for `hess` and `y_norm2` for `inv_hess` (looking formally).

But I guess testing for both `ys == 0 or y_norm2 == 0` is better, because making either matrix infinite or zero is not a good thing. What do you think?

Reply:
Ok, just rewrote it as suggested and included the test for ``y_norm2 == 0``. Thank you.

In theory, the test for ``y_norm2 == 0`` would not be necessary because, if ``y_norm2 == 0`` than ``y ==0``, which automatically implies that ``ys==0``. But its better to add some redundancy and avoid possible round-off errors.